### ISSUE TITLE
R files packaged in the same JAR can't interact with each other

### ISSUE THREAD

**Issue by NRBPerdijk** [oracle/fastr#64]
Hello,

When creating a jar where Java (or another JVM language) is used to interface and run R files, the R files can't reference each other via the normal `source("filepath")` method.

I've created an example project that uses Maven to create a Jar of a Java application that has R code embedded: 
[Attempting to refer to other R files in a Jar](https://github.com/NRBPerdijk/fastRBug/tree/accessingRFilesInJar)

To reproduce the issue: 
* clone it
* run a mvn clean package 
* try to run the created jar using a GraalVM bin/java (error occurs in RC13 and RC14, other versions have not been tested)

You'll get a stacktrace somewhat similar to: 
```
[1] "Hello Directly"
Called otherRFunctionInJar!
Exception in thread "main" Error in file(filename, "r", encoding = encoding) :
  cannot open the connection
In addition: Warning message:
In file(filename, "r", encoding = encoding) :
  cannot open file 'file:/path/to/folder/fastRBug/target/fastRBug-1.0-SNAPSHOT.jar!/r/otherRFunctionInJar.R': No such file or directory
	at <R> file(internal:1:0)
	at <R> source(unknown.r:58:1911-1950)
	at <R> firstR(jar:file:/Users/npe21003/RND/fastRBug/target/fastRBug-1.0-SNAPSHOT.jar!/r/rFunctionThatCallsOtherRFunction.R:2:38-56)
	at com.oracle.truffle.polyglot.PolyglotFunction.apply(PolyglotFunction.java:66)
	at Main.main(Main.java:16)
```
Removing `file:` from the provided path doesn't help, nor does hardcoding it in the `R` file that is sourcing the other `R` file.  

We've implemented a workaround that I will explain in a reply to this issue.

**Comment by NRBPerdijk** [comment]
The work around we used is using a maven plugin to move the R files outside of the jar into their own folder during packaging and then zipping the jar and the r folder together. This requires anyone who wants to run the application to unpack the resulting zip, before running the jar. The JVM code then references this R folder by its relative path and the R code can safely do the same. This is a bit of a hack, but it works for now.

**Comment by steve-s** [comment]
Hello Nathan, 

the `source` function is a tricky thing even if you do not want to use resources since it looks up the file relative to the current working directory and not relative to the file that invokes it. It is always trying to lookup a regular file on the filesystem. Anyway, that doesn't solve your problem.

There are several possible approaches I could think of:

- read the resource file on the Java/Scala side and pass it as a `String` to FastR, then in the R context evaluate it with something like `source(exprs = parse(text = content))`. (I've tested this, [diff.txt](https://github.com/oracle/fastr/files/3003758/diff.txt))


- create Java/Scala function that takes resource filename as `String` and evaluates it, e.g.:
```
boolean sourceResource(String path) {
   context.eval("R", Main.class.getResource(path));
}
```
and export that function (or object containing that function) to FastR and use it instead of `source`. (not tested)

- implement your own Truffle filesystem that delegates to the regular filesystem, but handles some special paths by looking them up in the resources. This way you could actually also package additional R packages as resources in your jar. Unfortunately, there is not a lot of documentation on Truffle filesystem SPI, but there is this example: https://github.com/tzezula/jimfs

**Comment by NRBPerdijk** [comment]
Hello Steve,

Some interesting suggestions! 
* The first feels a bit too messy: I expect the R side of a JVM/R project to "manage its own dependencies", so I want to do most of the heavy lifting on that side rather than straight up injecting source code as a String.
* I've created a working example of your second suggestion and pushed it [to the same branch](https://github.com/NRBPerdijk/fastRBug/tree/accessingRFilesInJar), so that if someone else stumbles across this issue report they can find that solution easily. I like it better than the maven packaging workaround I proposed, as there are fewer moving parts.
* The Truffle filesystem fix sounds interesting as a more durable solution. The desire to run a graal project as a jar will probably not be uncommon. Since package management is also a bit of a hairy issue in R, it could help deal with that as well. Due to time constraints, it is a bit too ambitious at this time for me to pursue though.
